src/Middleware/HeaderPropagation/src/DependencyInjection/HeaderPropagationExtensions.cs

```diff
@@ -0,0 +1,95 @@
+        {
+            if (services == null)
+            {
+                throw new ArgumentNullException(nameof(services));
+            }
+
+            services.TryAddSingleton<HeaderPropagationValues>();
+
+            return services;
+        }
+
+        /// <summary>
+        /// Adds services required for propagating headers to a <see cref="HttpClient"/>.
+        /// </summary>
+        /// <param name="services">The <see cref="IServiceCollection"/> to add the services to.</param>
+        /// <param name="configureOptions">The <see cref="HeaderPropagationOptions"/> to configure the middleware with.</param>
+        /// <returns>The <see cref="IServiceCollection"/> so that additional calls can be chained.</returns>
+        public static IServiceCollection AddHeaderPropagation(this IServiceCollection services, Action<HeaderPropagationOptions> configureOptions)
+        {
+            if (services == null)
+            {
+                throw new ArgumentNullException(nameof(services));
+            }
+
+            if (configureOptions == null)
+            {
+                throw new ArgumentNullException(nameof(configureOptions));
+            }
+
+            services.Configure(configureOptions);
+            services.AddHeaderPropagation();
+
+            return services;
+        }
+
+        /// <summary>
+        /// Adds a message handler for propagating headers collected by the <see cref="HeaderPropagationMiddleware"/> to a outgoing request.
+        /// </summary>
+        /// <param name="builder">The <see cref="IHttpClientBuilder"/> to add the message handler to.</param>
+        /// <returns>The <see cref="IHttpClientBuilder"/> so that additional calls can be chained.</returns>
+        public static IHttpClientBuilder AddHeaderPropagation(this IHttpClientBuilder builder)
+        {
+            builder.Services.TryAddSingleton<HeaderPropagationValues>();
+            builder.Services.TryAddTransient<HeaderPropagationMessageHandler>();
+
+            builder.AddHttpMessageHandler<HeaderPropagationMessageHandler>();
+
+            return builder;
+        }
+
+        /// <summary>
+        /// Adds a middleware that collect headers to be propagated to a <see cref="HttpClient"/>.
+        /// </summary>
+        /// <param name="app">The <see cref="IApplicationBuilder"/> to add the middleware to.</param>
+        /// <returns>A reference to the <paramref name="app"/> after the operation has completed.</returns>
+        public static IApplicationBuilder UseHeaderPropagation(this IApplicationBuilder app)
+        {
+            if (app == null)
+            {
+                throw new ArgumentNullException(nameof(app));
+            }
+
+            if (app.ApplicationServices.GetService<HeaderPropagationValues>() == null)
+            {
+                throw new InvalidOperationException(_unableToFindServices);
+            }
+
+            return app.UseMiddleware<HeaderPropagationMiddleware>();
+        }
+    }
```

Comment:
These methods should be split up into separate classes. One reason for this is that they need to go into different namespaces. We want `Add` and `Use` methods to be in scope by default in startup code.


I would expect to see the following:

`Microsoft.Extensions.DependencyInjection.HeaderPropagationServiceCollectionExtensions` -> `AddHeaderPropagation`
`Microsoft.Extensions.DependencyInjection.HeaderPropagationHttpClientBuilderExtensions` -> `AddHeaderPropagation`
`Microsoft.AspNetCore.Builder.HeaderPropagationApplicationBuilderExtensions` -> `UseHeaderPropagation`

Reply:
Makes sense